kratos/utilities/constraint_utilities.cpp

```diff
@@ -88,5 +88,128 @@ void ApplyConstraints(ModelPart& rModelPart)
+            const VariableComponentType& r_check_dof_z = KratosComponents<VariableComponentType>::Get(r_dof_variable_name + "_Z");
+
+            // Getting the residual dofs
+            const VariableComponentType& r_residual_dof_x = KratosComponents<VariableComponentType>::Get(r_reaction_variable_name + "_X");
+            const VariableComponentType& r_residual_dof_y = KratosComponents<VariableComponentType>::Get(r_reaction_variable_name + "_Y");
+            const VariableComponentType& r_residual_dof_z = KratosComponents<VariableComponentType>::Get(r_reaction_variable_name + "_Z");
+
+            components_variable_map.insert(std::pair<std::size_t, VariableComponentType>(r_check_dof_x.Key(), r_residual_dof_x));
+            components_variable_map.insert(std::pair<std::size_t, VariableComponentType>(r_check_dof_y.Key(), r_residual_dof_y));
+            components_variable_map.insert(std::pair<std::size_t, VariableComponentType>(r_check_dof_z.Key(), r_residual_dof_z));
+        } else {
+            KRATOS_ERROR << "Variable is not an array or a double" << std::endl;
+        }
+
+        ++counter;
+    }
+
+    // Getting auxiliar variables
+    const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+    const int number_of_constraints = static_cast<int>(rModelPart.MasterSlaveConstraints().size());
+    const auto it_cont_begin = rModelPart.MasterSlaveConstraints().begin();
+
+    // Auxiliar values
+    Matrix transformation_matrix;
+    Vector constant_vector, slave_solution_vector, master_solution_vector;
+    ModelPart::NodeType::Pointer p_master_node = *(rModelPart.Nodes().begin()).base();
+    ModelPart::NodeType::Pointer p_slave_node = *(rModelPart.Nodes().begin()).base();
+
+    #pragma omp parallel firstprivate(transformation_matrix, constant_vector, slave_solution_vector, master_solution_vector, p_master_node, p_slave_node)
+    {
+        #pragma omp for schedule(guided, 512)
+        for (int i_const = 0; i_const < number_of_constraints; ++i_const) {
+            auto it_const = it_cont_begin + i_const;
+
+            // Getting the transformation matrix and constant vector
+            it_const->CalculateLocalSystem(transformation_matrix, constant_vector, r_current_process_info);
+
+            // Resizing vectors
+            if (slave_solution_vector.size() != transformation_matrix.size1())
+                slave_solution_vector.resize(transformation_matrix.size1());
+            if (master_solution_vector.size() != transformation_matrix.size2())
+                master_solution_vector.resize(transformation_matrix.size2());
+
+            std::size_t counter = 0;
+            for (auto& r_dof_slave: it_const->GetSlaveDofsVector()) {
+                const std::size_t slave_variable_key = r_dof_slave->GetVariable().Key();
+
+                // Updating pointer
+                if (r_dof_slave->Id() != p_slave_node->Id())
+                    p_slave_node = rModelPart.pGetNode(r_dof_slave->Id());
```

Comment:
just to add that those searches might be slow in large models

Reply:
I am avoiding them as much as I can, the MPC are related to dofs, and dofs and nodes are strongly tied. Variables stored in the nodes are not necessarily dofs, so you need to call for the node. AN alternative solution would be add the residual forces as dofs..., so you don't need the nodes